advanced-cache.php

```diff
@@ -551,11 +551,14 @@ function add_debug_html_to_output( $debug_html ) {
 
 // Recreate the permalink from the URL
 $batcache->permalink = 'http://' . $batcache->keys['host'] . $batcache->keys['path'] . ( isset($batcache->keys['query']['p']) ? "?p=" . $batcache->keys['query']['p'] : '' );
-$batcache->url_key = md5($batcache->permalink);
+$batcache->generate_keys();
```

Comment:
There's a reason this appears at the end - `$batcache->do_variants()` add extra keys that would alter the value of `$batcache->key`. Right now this would break anything using `vary_cache_on_function` so the solution isn't quite so straightforward.

`$batcache->do_variants()` requires `$batcache->url_key` to be set, so I guess that's why they're in this order at the moment. Looks like the need for 2 cache keys is purely to support the way variants are handled. It might work to just call `$batcache->generate_keys()` again after `$batcache->do_variants()`.

Reply:
Seems tied into the cache `version` value too, that's used by the `batcache_clear_url()` function which could be used all sorts of ways by folks. You'd expect to be able to pass the base URL without any query strings and flush all matching URLs.

Reply:
Ah ok, so I think in that case, it might be better to just change the genlock caches specifically, rather than changing "url_key". That should still fix the case I'm wanting to (store a genlock per batcache key, rather than per url)?

Reply:
Yeah `{$batcache->url_key}_genlock` is only referenced in the output buffer callback and after `$batcache->generate_keys()` is called. Within the `ob()` callback `$batcache->do_variants($this->vary)` and `$batcache->generate_keys()` are called again but I think this is actually fine in the case of genlock.

Yup, changing all instances of `{$batcache->url_key}_genlock` to `{$batcache->key}_genlock` should be a goer.

Reply:
@roborourke ok updated!